File: src/android/ansight-core/src/main/kotlin/ai/ansight/runtime/AnsightModels.kt
package ai.ansight.runtime

import java.util.UUID
import kotlin.math.ceil

object AnsightSamplingLimits {
    const val DefaultSampleFrequencyMilliseconds = 500
    const val MinSampleFrequencyMilliseconds = 200
    const val MaxSampleFrequencyMilliseconds = 2_000
    const val DefaultRetentionPeriodSeconds = 600
    const val MinRetentionPeriodSeconds = 60
    const val MaxRetentionPeriodSeconds = 3_600
}

object AnsightChannels {
    const val JavaHeap = 0
    const val NativeHeap = 1
    const val Rss = 2
    const val FramesPerSecond = 3
    const val Lifecycle = 4
    const val BatteryLevel = 5
    const val JniReferenceCount = 6
    const val OpenFileHandles = 7
    const val Unspecified = 255

    val reservedIds: Set<Int> = setOf(
        JavaHeap,
        NativeHeap,
        Rss,
        FramesPerSecond,
        Lifecycle,
        BatteryLevel,
        JniReferenceCount,
        OpenFileHandles,
        Unspecified,
    )
}

data class AnsightChannel(
    val id: Int,
    val name: String,
    val colorHex: String? = null,
    val unit: String? = null,
    val type: String = "custom",
    val source: String? = null,
    val group: String? = null,
    val kind: String? = null,
) {
    fun validated(): AnsightChannel {
        require(id in 0..255) { "Channel ids must be between 0 and 255." }
        require(name.isNotBlank()) { "Channel names must not be blank." }
        return copy(
            name = name.trim(),
            colorHex = colorHex?.trim()?.ifBlank { null },
            unit = unit?.trim()?.ifBlank { null },
            type = type.trim().ifBlank { "custom" },
            source = source?.trim()?.ifBlank { null },
            group = group?.trim()?.ifBlank { null },
            kind = kind?.trim()?.ifBlank { null },
        )
    }
}

data class DefaultMemoryChannels(
    val javaHeap: Boolean = true,
    val nativeHeap: Boolean = true,
    val rss: Boolean = true,
) {
    companion object {
        val PlatformDefaults = DefaultMemoryChannels()
        val None = DefaultMemoryChannels(javaHeap = false, nativeHeap = false, rss = false)
    }
}

enum class AnsightSessionJpegCaptureMode {
    ScreenshotOnly,
    ScreenshotAndVisualTree,
}

data class AnsightSessionJpegCaptureOptions(
    val intervalMilliseconds: Int = DefaultIntervalMilliseconds,
    val quality: Int = DefaultQuality,
    val maxWidth: Int? = DefaultMaxWidth,
    val captureGpuBackedSurfaces: Boolean = DefaultCaptureGpuBackedSurfaces,
    val mode: AnsightSessionJpegCaptureMode = DefaultMode,
) {
    companion object {
        const val DefaultIntervalMilliseconds = 2_000
        const val DefaultQuality = 60
        const val DefaultMaxWidth = 480
        const val DefaultCaptureGpuBackedSurfaces = true
        val DefaultMode = AnsightSessionJpegCaptureMode.ScreenshotOnly
    }

    fun validated(): AnsightSessionJpegCaptureOptions {
        return copy(
            intervalMilliseconds = intervalMilliseconds.coerceAtLeast(250),
            quality = quality.coerceIn(1, 100),
            maxWidth = maxWidth?.takeIf { it > 0 }?.coerceAtMost(8_192),
        )
    }
}

data class AnsightTouchCaptureOptions(
    val moveCaptureDistanceThreshold: Double = 8.0,
    val moveCaptureFramesPerSecond: Int = 20,
) {
    fun validated(): AnsightTouchCaptureOptions {
        return copy(
            moveCaptureDistanceThreshold = moveCaptureDistanceThreshold.takeIf { it.isFinite() && it >= 0.0 } ?: 8.0,
            moveCaptureFramesPerSecond = moveCaptureFramesPerSecond.coerceAtLeast(0),
        )
    }
}

enum class AnsightToolGuard {
    Disabled,
    ReadOnly,
    ReadWrite,
    FullAccess,
}

data class AnsightHostAutoProbeOptions(
    val enabled: Boolean = true,
    val initialDelayMilliseconds: Long = 1_000,
    val probeIntervalMilliseconds: Long = 5_000,
    val reconnectDelayMilliseconds: Long = 10_000,
    val clientName: String? = null,
) {
    fun validated(): AnsightHostAutoProbeOptions {
        return copy(
            initialDelayMilliseconds = initialDelayMilliseconds.coerceAtLeast(0),
            probeIntervalMilliseconds = probeIntervalMilliseconds.coerceAtLeast(1_000),
            reconnectDelayMilliseconds = reconnectDelayMilliseconds.coerceAtLeast(1_000),
            clientName = clientName?.trim()?.ifBlank { null },
        )
    }
}

data class AnsightHostConnectionOptions(
    val savedConfigKey: String = "ai.ansight.android.saved-pairing",
    val bundledConfigJson: String? = null,
    val discoveryPort: Int? = null,
    val allowCellularConnections: Boolean = false,
    val connectionProfileRetentionSeconds: Long = 14L * 24L * 60L * 60L,
    val configReader: HostConnectionConfigReader? = null,
) {
    fun validated(): AnsightHostConnectionOptions {
        if (discoveryPort != null) {
            require(discoveryPort in 1..65_535) { "Discovery port must be between 1 and 65535." }
        }

        return copy(
            savedConfigKey = savedConfigKey.trim().ifBlank { "ai.ansight.android.saved-pairing" },
            bundledConfigJson = bundledConfigJson?.trim()?.ifBlank { null },
            connectionProfileRetentionSeconds = connectionProfileRetentionSeconds.coerceAtLeast(1),
        )
    }
}

interface HostConnectionConfigReader {
    fun canRead(kind: HostConnectionRequestKind): Boolean
    fun readConfigPayload(request: HostConnectionRequest): String?
}

data class AnsightSecureStorageOptions(
    val preferencesName: String = "ai.ansight.secure-storage",
    val allowedKeys: Set<String> = emptySet(),
    val allowedPrefixes: Set<String> = emptySet(),
) {
    fun validated(): AnsightSecureStorageOptions {
        return copy(
            preferencesName = preferencesName.trim().ifBlank { "ai.ansight.secure-storage" },
            allowedKeys = allowedKeys.mapNotNull { it.trim().ifBlank { null } }.toSet(),
            allowedPrefixes = allowedPrefixes.mapNotNull { it.trim().ifBlank { null } }.toSet(),
        )
    }

    fun isAllowed(key: String): Boolean {
        val normalized = key.trim()
        return normalized in allowedKeys || allowedPrefixes.any { normalized.startsWith(it) }
    }
}

data class AnsightOptions(
    val sampleFrequencyMilliseconds: Int = AnsightSamplingLimits.DefaultSampleFrequencyMilliseconds,
    val retentionPeriodSeconds: Int = AnsightSamplingLimits.DefaultRetentionPeriodSeconds,
    val enableFramesPerSecond: Boolean = true,
    val enableBatteryLevel: Boolean = false,
    val enableOpenFileHandleTracking: Boolean = false,
    val enableJniReferenceCountTracking: Boolean = false,
    val additionalChannels: List<AnsightChannel> = emptyList(),
    val defaultMemoryChannels: DefaultMemoryChannels = DefaultMemoryChannels.PlatformDefaults,
    val sessionJpegCapture: AnsightSessionJpegCaptureOptions? = null,
    val touchCapture: AnsightTouchCaptureOptions? = AnsightTouchCaptureOptions(),
    val toolGuard: AnsightToolGuard = AnsightToolGuard.Disabled,
    val customProperties: Map<String, Map<String, String>> = emptyMap(),
    val hostAutoProbe: AnsightHostAutoProbeOptions = AnsightHostAutoProbeOptions(),
    val hostConnection: AnsightHostConnectionOptions = AnsightHostConnectionOptions(),
    val crashCapture: AnsightCrashCaptureOptions = AnsightCrashCaptureOptions(),
    val secureStorage: AnsightSecureStorageOptions = AnsightSecureStorageOptions(),
    val initialTools: List<AndroidTool> = emptyList(),
    val artifactProviders: List<AndroidArtifactProvider> = emptyList(),
) {
    companion object {
        @JvmStatic
        fun createBuilder(): AnsightOptionsBuilder = AnsightOptionsBuilder()

        @JvmStatic
        fun createBuilder(options: AnsightOptions): AnsightOptionsBuilder = AnsightOptionsBuilder(options)
    }

    val maximumBufferSize: Int
        get() = retentionPeriodSeconds * ceil(1000.0 / sampleFrequencyMilliseconds.toDouble()).toInt()

    fun validated(): AnsightOptions {
        val validatedChannels = additionalChannels.map { channel ->
            val validated = channel.validated()
            require(validated.id !in AnsightChannels.reservedIds) {
                "Additional channel '${validated.name}' uses reserved channel id ${validated.id}."
            }
            validated
        }

        return copy(
            sampleFrequencyMilliseconds = sampleFrequencyMilliseconds.coerceIn(
                AnsightSamplingLimits.MinSampleFrequencyMilliseconds,
                AnsightSamplingLimits.MaxSampleFrequencyMilliseconds,
            ),
            retentionPeriodSeconds = retentionPeriodSeconds.coerceIn(
                AnsightSamplingLimits.MinRetentionPeriodSeconds,
                AnsightSamplingLimits.MaxRetentionPeriodSeconds,
            ),
            additionalChannels = validatedChannels,
            sessionJpegCapture = sessionJpegCapture?.validated(),
            touchCapture = touchCapture?.validated(),
            customProperties = customProperties.normalizedCustomProperties(),
            hostAutoProbe = hostAutoProbe.validated(),
            hostConnection = hostConnection.validated(),
            crashCapture = crashCapture.validated(),
            secureStorage = secureStorage.validated(),
        )
    }
}

enum class AnsightEventType(val wireName: String) {
    Event("Event"),
    Debug("Debug"),
    Info("Info"),
    Warning("Warning"),
    Error("Error"),
    Exception("Exception"),
    Gc("Gc"),
    Navigation("Navigation"),
    ScreenViewed("ScreenViewed"),
    Lifecycle("Lifecycle"),
}

enum class AppLifecycleState(val wireName: String) {
    Unknown("unknown"),
    Foreground("foreground"),
    Background("background"),
}

enum class HostConnectionRequestKind {
    Auto,
    SavedConfig,
    BundledConfig,
    File,
    QrCode,
    Payload,
    Config,
}

enum class HostConnectionSource {
    None,
    AutoProbe,
    CachedSession,
    SavedConfig,
    BundledConfig,
    Payload,
    ConfigReader,
    HostConnection,
    Transport,
    Telemetry,
    AppState,
    SessionJpegCapture,
    TouchCapture,
}

enum class HostConnectionState {
    Disconnected,
    Connecting,
    Connected,
    Disconnecting,
    Failed,
    Unavailable,
}

enum class HostConnectionSummaryKind {
    None,
    Ready,
    Connected,
    Disconnected,
    Failed,
    Unavailable,
}

enum class HostConnectionActionKind {
    None,
    Connect,
    Disconnect,
    ClearSavedConfig,
    NotifyConfigChanged,
}

data class HostConnectionRequest(
    val kind: HostConnectionRequestKind = HostConnectionRequestKind.Auto,
    val payload: String? = null,
    val clientName: String? = null,
    val expectedAppId: String? = null,
    val hostAddressOverride: String? = null,
) {
    companion object {
        @JvmStatic
        fun auto(
            clientName: String? = null,
            expectedAppId: String? = null,
            hostAddressOverride: String? = null,
        ) = HostConnectionRequest(
            kind = HostConnectionRequestKind.Auto,
            clientName = clientName,
            expectedAppId = expectedAppId,
            hostAddressOverride = hostAddressOverride,
        )

        @JvmStatic
        fun savedConfig(
            clientName: String? = null,
            expectedAppId: String? = null,
            hostAddressOverride: String? = null,
        ) = HostConnectionRequest(
            kind = HostConnectionRequestKind.SavedConfig,
            clientName = clientName,
            expectedAppId = expectedAppId,
            hostAddressOverride = hostAddressOverride,
        )

        @JvmStatic
        fun bundledConfig(
            clientName: String? = null,
            expectedAppId: String? = null,
            hostAddressOverride: String? = null,
        ) = HostConnectionRequest(
            kind = HostConnectionRequestKind.BundledConfig,
            clientName = clientName,
            expectedAppId = expectedAppId,
            hostAddressOverride = hostAddressOverride,
        )

        @JvmStatic
        fun file(
            path: String? = null,
            clientName: String? = null,
            expectedAppId: String? = null,
            hostAddressOverride: String? = null,
        ) = HostConnectionRequest(
            kind = HostConnectionRequestKind.File,
            payload = path,
            clientName = clientName,
            expectedAppId = expectedAppId,
            hostAddressOverride = hostAddressOverride,
        )

        @JvmStatic
        fun qrCode(
            clientName: String? = null,
            expectedAppId: String? = null,
            hostAddressOverride: String? = null,
        ) = HostConnectionRequest(
            kind = HostConnectionRequestKind.QrCode,
            clientName = clientName,
            expectedAppId = expectedAppId,
            hostAddressOverride = hostAddressOverride,
        )

        @JvmStatic
        fun payloadText(
            payload: String,
            clientName: String? = null,
            expectedAppId: String? = null,
            hostAddressOverride: String? = null,
        ) = HostConnectionRequest(
            kind = HostConnectionRequestKind.Payload,
            payload = payload,
            clientName = clientName,
            expectedAppId = expectedAppId,
            hostAddressOverride = hostAddressOverride,
        )
    }
}

data class HostConnectionStatus(
    val isRuntimeActive: Boolean,
    val isConnected: Boolean,
    val connectionState: HostConnectionState,
    val hasCachedSession: Boolean,
    val hasSavedConfig: Boolean,
    val hasBundledConfig: Boolean,
    val summaryKind: HostConnectionSummaryKind,
    val summaryMessage: String,
)

data class HostConnectionCapabilities(
    val canConnectUsingSavedConfig: Boolean,
    val canConnectUsingBundledConfig: Boolean,
    val canChooseConfigFile: Boolean,
    val canScanConfigQrCode: Boolean,
    val canClearSavedConfigs: Boolean,
)

fun interface HostConnectionStatusListener {
    fun onChanged(status: HostConnectionStatus, capabilities: HostConnectionCapabilities)
}

class HostConnectionStatusSubscription internal constructor(
    private val removeAction: () -> Unit,
) : AutoCloseable {
    private val removed = java.util.concurrent.atomic.AtomicBoolean(false)

    fun remove() {
        if (removed.compareAndSet(false, true)) {
            removeAction()
        }
    }

    override fun close() {
        remove()
    }
}

data class HostConnectionResult(
    val success: Boolean,
    val message: String,
    val kind: HostConnectionActionKind = HostConnectionActionKind.None,
    val source: HostConnectionSource = HostConnectionSource.None,
    val reasonCode: String? = null,
    val openSession: OpenSessionResult? = null,
) {
    companion object {
        fun success(
            message: String,
            kind: HostConnectionActionKind = HostConnectionActionKind.None,
            source: HostConnectionSource = HostConnectionSource.None,
            openSession: OpenSessionResult? = null,
        ) = HostConnectionResult(true, message, kind, source, openSession = openSession)

        fun failure(
            message: String,
            kind: HostConnectionActionKind = HostConnectionActionKind.None,
            source: HostConnectionSource = HostConnectionSource.None,
            reasonCode: String? = null,
            openSession: OpenSessionResult? = null,
        ) = HostConnectionResult(false, message, kind, source, reasonCode, openSession)
    }
}

data class PairingOpenOptions(
    val clientName: String,
    val expectedAppId: String? = null,
    val hostAddressOverride: String? = null,
    val profileOverride: Map<String, String> = emptyMap(),
)

data class OpenSessionResult(
    val success: Boolean,
    val message: String,
    val accepted: Boolean = success,
    val sessionId: String? = null,
    val configId: String? = null,
    val appId: String? = null,
    val resolvedHostAddress: String? = null,
    val discoverySource: String? = null,
    val reasonCode: String? = null,
    val hostId: String? = null,
    val hostName: String? = null,
)

data class OperationResult(
    val success: Boolean,
    val message: String,
) {
    companion object {
        fun success(message: String) = OperationResult(true, message)
        fun failure(message: String) = OperationResult(false, message)
    }
}

data class AnsightToolDescriptor(
    val id: String,
    val name: String,
    val scope: String = "Read",
)

data class RecordedMetric(
    val value: Long,
    val channel: Int,
    val capturedAtUtc: String = AnsightClock.isoNow(),
    val capturedAtEpochMs: Long = System.currentTimeMillis(),
    val sequence: Long = 0,
)

data class RecordedEvent(
    val id: String = UUID.randomUUID().toString(),
    val label: String,
    val type: AnsightEventType,
    val details: String?,
    val channel: Int,
    val capturedAtUtc: String = AnsightClock.isoNow(),
    val capturedAtEpochMs: Long = System.currentTimeMillis(),
    val externalId: String? = null,
    val sequence: Long = 0,
)

data class RecordedScreenView(
    val name: String,
    val details: Map<String, String> = emptyMap(),
    val capturedAtUtc: String = AnsightClock.isoNow(),
)

data class RecordedTouch(
    val id: String,
    val action: String,
    val pointerId: Long,
    val pointerIndex: Int,
    val pointerCount: Int,
    val x: Double,
    val y: Double,
    val surfaceWidth: Double?,
    val surfaceHeight: Double?,
    val coordinateUnit: String,
    val surfaceScale: Double?,
    val normalizedX: Double?,
    val normalizedY: Double?,
    val capturedAtUtc: String,
    val capturedAtEpochMs: Long,
    val sequence: Long,
)

data class AnsightDebugSnapshot(
    val initialized: Boolean,
    val active: Boolean,
    val sessionOpen: Boolean,
    val connectionStatus: HostConnectionStatus,
    val lifecycleState: AppLifecycleState,
    val lifecycleChangedAtUtc: String?,
    val currentScreen: RecordedScreenView?,
    val metricsRecorded: Int,
    val eventsRecorded: Int,
    val touchesRecorded: Int,
    val channels: List<AnsightChannel>,
    val registeredTools: Int,
    val lastMetric: RecordedMetric?,
    val lastEvent: RecordedEvent?,
    val deviceProfile: DeviceAppProfile?,
    val sessionMessage: String?,
)

object PairingFailureCodes {
    const val HostAddressRequired = "HostAddressRequired"
    const val WifiRequired = "WifiRequired"
    const val EnrollmentRequired = "EnrollmentRequired"
    const val EnrollmentUnavailable = "EnrollmentUnavailable"
    const val EnrollmentExpired = "EnrollmentExpired"
    const val EnrollmentConsumed = "EnrollmentConsumed"
    const val AccessTokenInvalid = "AccessTokenInvalid"
    const val RegistrationExpired = "RegistrationExpired"
    const val SignInRequired = "SignInRequired"
    const val UdpBootstrapFailed = "UdpBootstrapFailed"
    const val UdpBootstrapTimeout = "UdpBootstrapTimeout"
    const val WebSocketHandoffUnavailable = "WebSocketHandoffUnavailable"
    const val WebSocketEndpointUnreachable = "WebSocketEndpointUnreachable"
    const val WebSocketHandshakeFailed = "WebSocketHandshakeFailed"
}

File: src/flutter/lib/src/ansight_models.dart
typedef AnsightJson = Map<String, Object?>;

enum AnsightEventType {
  event('Event'),
  debug('Debug'),
  info('Info'),
  warning('Warning'),
  error('Error'),
  exception('Exception'),
  gc('Gc'),
  navigation('Navigation'),
  screenViewed('ScreenViewed'),
  lifecycle('Lifecycle');

  const AnsightEventType(this.wireName);

  final String wireName;
}

enum AnsightLifecycleState {
  unknown('unknown'),
  foreground('foreground'),
  background('background');

  const AnsightLifecycleState(this.wireName);

  final String wireName;
}

enum AnsightToolGuard {
  disabled('disabled'),
  readOnly('readOnly'),
  readWrite('readWrite'),
  fullAccess('fullAccess');

  const AnsightToolGuard(this.wireName);

  final String wireName;
}

enum AnsightToolScope {
  read('read'),
  write('write'),
  delete('delete');

  const AnsightToolScope(this.wireName);

  final String wireName;
}

enum AnsightToolSecurityLevel {
  unspecified('unspecified'),
  low('low'),
  medium('medium'),
  moderate('moderate'),
  high('high'),
  critical('critical');

  const AnsightToolSecurityLevel(this.wireName);

  final String wireName;
}

class AnsightChannel {
  const AnsightChannel({
    required this.id,
    required this.name,
    this.unit,
    this.type = 'custom',
    this.colorHex,
    this.source,
    this.group,
    this.kind,
  });

  final int id;
  final String name;
  final String? unit;
  final String type;
  final String? colorHex;
  final String? source;
  final String? group;
  final String? kind;

  AnsightJson toJson() => <String, Object?>{
        'id': id,
        'name': name,
        if (unit != null) 'unit': unit,
        'type': type,
        if (colorHex != null) 'colorHex': colorHex,
        if (source != null) 'source': source,
        if (group != null) 'group': group,
        if (kind != null) 'kind': kind,
      };

  factory AnsightChannel.fromJson(AnsightJson json) => AnsightChannel(
        id: _int(json['id']),
        name: _string(json['name']),
        unit: _nullableString(json['unit']),
        type: _nullableString(json['type']) ?? 'custom',
        colorHex: _nullableString(json['colorHex']),
        source: _nullableString(json['source']),
        group: _nullableString(json['group']),
        kind: _nullableString(json['kind']),
      );
}

class AnsightOperationResult {
  const AnsightOperationResult({
    required this.success,
    required this.message,
    this.errorCode,
    this.data = const <String, Object?>{},
  });

  final bool success;
  final String message;
  final String? errorCode;
  final AnsightJson data;

  factory AnsightOperationResult.fromJson(AnsightJson json) =>
      AnsightOperationResult(
        success: _bool(json['success']),
        message: _nullableString(json['message']) ?? '',
        errorCode: _nullableString(json['errorCode']),
        data: Map<String, Object?>.unmodifiable(json),
      );
}

class AnsightHostConnectionResult {
  const AnsightHostConnectionResult({
    required this.success,
    required this.message,
    this.kind,
    this.source,
    this.reasonCode,
    this.accepted,
    this.sessionId,
    this.configId,
    this.appId,
    this.resolvedHostAddress,
    this.usedEmbeddedDeveloperPairing,
    this.discoverySource,
    this.hostId,
    this.hostName,
    this.data = const <String, Object?>{},
  });

  final bool success;
  final String message;
  final String? kind;
  final String? source;
  final String? reasonCode;
  final bool? accepted;
  final String? sessionId;
  final String? configId;
  final String? appId;
  final String? resolvedHostAddress;
  final bool? usedEmbeddedDeveloperPairing;
  final String? discoverySource;
  final String? hostId;
  final String? hostName;
  final AnsightJson data;

  factory AnsightHostConnectionResult.fromJson(AnsightJson json) =>
      AnsightHostConnectionResult(
        success: _bool(json['success']),
        message: _nullableString(json['message']) ?? '',
        kind: _nullableString(json['kind']),
        source: _nullableString(json['source']),
        reasonCode: _nullableString(json['reasonCode']),
        accepted: _nullableBool(json['accepted']),
        sessionId: _nullableString(json['sessionId']),
        configId: _nullableString(json['configId']),
        appId: _nullableString(json['appId']),
        resolvedHostAddress: _nullableString(json['resolvedHostAddress']),
        usedEmbeddedDeveloperPairing: _nullableBool(
          json['usedEmbeddedDeveloperPairing'],
        ),
        discoverySource: _nullableString(json['discoverySource']),
        hostId: _nullableString(json['hostId']),
        hostName: _nullableString(json['hostName']),
        data: Map<String, Object?>.unmodifiable(json),
      );
}

class AnsightHostConnectionStatus {
  const AnsightHostConnectionStatus({
    required this.isRuntimeActive,
    required this.isConnected,
    required this.connectionState,
    required this.hasCachedSession,
    required this.hasSavedConfig,
    required this.hasBundledConfig,
    required this.summaryKind,
    required this.summaryMessage,
    this.data = const <String, Object?>{},
  });

  final bool isRuntimeActive;
  final bool isConnected;
  final String connectionState;
  final bool hasCachedSession;
  final bool hasSavedConfig;
  final bool hasBundledConfig;
  final String summaryKind;
  final String summaryMessage;
  final AnsightJson data;

  factory AnsightHostConnectionStatus.fromJson(AnsightJson json) =>
      AnsightHostConnectionStatus(
        isRuntimeActive: _bool(json['isRuntimeActive']),
        isConnected: _bool(json['isConnected']),
        connectionState: _nullableString(json['connectionState']) ?? 'unknown',
        hasCachedSession: _bool(json['hasCachedSession']),
        hasSavedConfig: _bool(json['hasSavedConfig']),
        hasBundledConfig: _bool(json['hasBundledConfig']),
        summaryKind: _nullableString(json['summaryKind']) ?? 'unknown',
        summaryMessage: _nullableString(json['summaryMessage']) ?? '',
        data: Map<String, Object?>.unmodifiable(json),
      );
}

class AnsightHostConnectionCapabilities {
  const AnsightHostConnectionCapabilities({
    required this.canConnectUsingSavedConfig,
    required this.canConnectUsingBundledConfig,
    required this.canChooseConfigFile,
    required this.canScanConfigQrCode,
    required this.canClearSavedConfigs,
    this.data = const <String, Object?>{},
  });

  final bool canConnectUsingSavedConfig;
  final bool canConnectUsingBundledConfig;
  final bool canChooseConfigFile;
  final bool canScanConfigQrCode;
  final bool canClearSavedConfigs;
  final AnsightJson data;

  factory AnsightHostConnectionCapabilities.fromJson(AnsightJson json) =>
      AnsightHostConnectionCapabilities(
        canConnectUsingSavedConfig: _bool(json['canConnectUsingSavedConfig']),
        canConnectUsingBundledConfig: _bool(
          json['canConnectUsingBundledConfig'],
        ),
        canChooseConfigFile: _bool(json['canChooseConfigFile']),
        canScanConfigQrCode: _bool(json['canScanConfigQrCode']),
        canClearSavedConfigs: _bool(json['canClearSavedConfigs']),
        data: Map<String, Object?>.unmodifiable(json),
      );
}

class AnsightRecordedMetric {
  const AnsightRecordedMetric({
    required this.value,
    required this.channel,
    this.recordedAtUtc,
    this.sequence,
    this.data = const <String, Object?>{},
  });

  final int value;
  final int channel;
  final DateTime? recordedAtUtc;
  final int? sequence;
  final AnsightJson data;

  factory AnsightRecordedMetric.fromJson(AnsightJson json) =>
      AnsightRecordedMetric(
        value: _int(json['value']),
        channel: _int(json['channel']),
        recordedAtUtc: _dateTime(json['recordedAtUtc'] ?? json['timestampUtc']),
        sequence: _nullableInt(json['sequence']),
        data: Map<String, Object?>.unmodifiable(json),
      );
}

class AnsightRecordedEvent {
  const AnsightRecordedEvent({
    required this.label,
    this.type,
    this.details,
    this.channel,
    this.recordedAtUtc,
    this.sequence,
    this.data = const <String, Object?>{},
  });

  final String label;
  final String? type;
  final String? details;
  final int? channel;
  final DateTime? recordedAtUtc;
  final int? sequence;
  final AnsightJson data;

  factory AnsightRecordedEvent.fromJson(AnsightJson json) =>
      AnsightRecordedEvent(
        label: _nullableString(json['label']) ?? '',
        type: _nullableString(json['type']),
        details: _nullableString(json['details']),
        channel: _nullableInt(json['channel']),
        recordedAtUtc: _dateTime(json['recordedAtUtc'] ?? json['timestampUtc']),
        sequence: _nullableInt(json['sequence']),
        data: Map<String, Object?>.unmodifiable(json),
      );
}

class AnsightDebugSnapshot {
  const AnsightDebugSnapshot({
    required this.initialized,
    required this.active,
    required this.sessionOpen,
    required this.metricsRecorded,
    required this.eventsRecorded,
    required this.registeredTools,
    required this.channels,
    required this.connectionStatus,
    this.lifecycleState,
    this.touchesRecorded,
    this.touchesCaptured,
    this.touchesSent,
    this.executableTools,
    this.sessionMessage,
    this.lastMetric,
    this.lastEvent,
    this.data = const <String, Object?>{},
  });

  final bool initialized;
  final bool active;
  final bool sessionOpen;
  final String? lifecycleState;
  final int metricsRecorded;
  final int eventsRecorded;
  final int? touchesRecorded;
  final int? touchesCaptured;
  final int? touchesSent;
  final int registeredTools;
  final int? executableTools;
  final String? sessionMessage;
  final List<AnsightChannel> channels;
  final AnsightHostConnectionStatus connectionStatus;
  final AnsightRecordedMetric? lastMetric;
  final AnsightRecordedEvent? lastEvent;
  final AnsightJson data;

  factory AnsightDebugSnapshot.fromJson(AnsightJson json) {
    final connection = _jsonMap(json['connectionStatus']);
    return AnsightDebugSnapshot(
      initialized: _bool(json['initialized']),
      active: _bool(json['active']),
      sessionOpen: _bool(json['sessionOpen']),
      lifecycleState: _nullableString(json['lifecycleState']),
      metricsRecorded: _int(json['metricsRecorded']),
      eventsRecorded: _int(json['eventsRecorded']),
      touchesRecorded: _nullableInt(json['touchesRecorded']),
      touchesCaptured: _nullableInt(json['touchesCaptured']),
      touchesSent: _nullableInt(json['touchesSent']),
      registeredTools: _int(json['registeredTools']),
      executableTools: _nullableInt(json['executableTools']),
      sessionMessage: _nullableString(json['sessionMessage']),
      channels: _jsonList(
        json['channels'],
      ).map(_jsonMap).map(AnsightChannel.fromJson).toList(growable: false),
      connectionStatus: AnsightHostConnectionStatus.fromJson(connection),
      lastMetric: json['lastMetric'] == null
          ? null
          : AnsightRecordedMetric.fromJson(_jsonMap(json['lastMetric'])),
      lastEvent: json['lastEvent'] == null
          ? null
          : AnsightRecordedEvent.fromJson(_jsonMap(json['lastEvent'])),
      data: Map<String, Object?>.unmodifiable(json),
    );
  }
}

class AnsightLogEntry {
  const AnsightLogEntry({
    required this.level,
    required this.message,
    this.platform,
    this.error,
    this.data = const <String, Object?>{},
  });

  final String level;
  final String message;
  final String? platform;
  final String? error;
  final AnsightJson data;

  factory AnsightLogEntry.fromJson(AnsightJson json) => AnsightLogEntry(
        level: _nullableString(json['level']) ?? 'info',
        message: _nullableString(json['message']) ?? '',
        platform: _nullableString(json['platform']),
        error: _nullableString(json['error']),
        data: Map<String, Object?>.unmodifiable(json),
      );
}

AnsightJson ansightJsonMap(Object? value) => _jsonMap(value);

List<Object?> ansightJsonList(Object? value) => _jsonList(value);

String _string(Object? value) => value?.toString() ?? '';

String? _nullableString(Object? value) {
  if (value == null) {
    return null;
  }
  return value.toString();
}

bool _bool(Object? value) => value is bool ? value : false;

bool? _nullableBool(Object? value) => value is bool ? value : null;

int _int(Object? value) => _nullableInt(value) ?? 0;

int? _nullableInt(Object? value) {
  if (value is int) {
    return value;
  }
  if (value is num) {
    return value.toInt();
  }
  return int.tryParse(value?.toString() ?? '');
}

DateTime? _dateTime(Object? value) =>
    DateTime.tryParse(value?.toString() ?? '');

AnsightJson _jsonMap(Object? value) {
  if (value is Map<String, Object?>) {
    return value;
  }
  if (value is Map) {
    return value.map(
      (Object? key, Object? item) => MapEntry(key.toString(), item),
    );
  }
  return <String, Object?>{};
}

List<Object?> _jsonList(Object? value) =>
    value is List ? List<Object?>.from(value) : const <Object?>[];
